### PR TITLE
Allow for presetting settings after a rename with /trn

### PR DESCRIPTION
**Pull request by scheibo**
Required for Zarel/Pokemon-Showdown-Client#1271

### CODE REVIEW

**Comment by scheibo** [comment]
The problem with this change is that an updated client (sending settings) talking to an old server (whose `/trn` method doesn't understand settings and thinks they're part of the token) isn't going to work. I could smuggle in settings as part of the second parameter (encode settings to it has no commas, and then `parseInt('0asfjhsajf')` still works on old servers etc), but that's pretty hideous. Is my only option here to add a new `/trn++` command, or is there a better way to negotiate with old servers/we don't care?

File: server/users.js [hunk 1]
```diff
@@ -544,6 +545,16 @@ class User extends Chat.MessageContext {
 		Users.add(this);
 	}
 
+	/**
+	 * @param {UserSettings} [settings]
+	 */
+	applySettings(settings) {
+		if (!settings) return;
+		if ('blockChallenges' in settings) this.blockChallenges = !!settings.blockChallenges;
```
Comment (TheJetOU) [inline server/users.js]: Why the `!!`? Won't the object be something like `{blockChallenges: true, ignoreTickets: false}`? In which case it's the same without the `!!`.

File: server/users.js [hunk 2]
```diff
@@ -544,6 +545,16 @@ class User extends Chat.MessageContext {
 		Users.add(this);
 	}
 
+	/**
+	 * @param {UserSettings} [settings]
```
Comment (TheJetOU) [inline server/users.js]: `settings` isn't an optional parameter though. You shouldn't be running this function with no parameters. It should probably be something like 
```suggestion
	 * @param {UserSettings?}
```

File: server/users.js (hunk 1 above)
Comment (scheibo) [reply server/users.js]: Because the compiler complains otherwise. I think its stupid as well.

**Comment by scheibo** [comment]
> Assertions can have commas in them which will ruin the parsing so you should probably move the settings before the assertion.

This is what I asked in `<<dev>>` - I figured that would probably be the case given we're not just doing a `target.split(',')`. 

---

The only way to move before the parsing and not break is to like base64 encode the JSON (or not use json and do a different scheme), but also to make it so that it can be `parseInt`-ed. I think a `/trns` alternative is probably the answer here

File: server/users.js (hunk 2 above)
Comment (scheibo) [reply server/users.js]: Ugh. Of course you're missing `settings` now

Comment (TheJetOU) [reply server/users.js]: my bad

File: server/chat-commands.js [hunk 3]
```diff
@@ -4488,17 +4488,24 @@ const commands = {
...
 			targetRegistered = target;
 			if (commaIndex >= 0) {
 				targetRegistered = !!parseInt(target.substr(0, commaIndex));
-				targetToken = target.substr(commaIndex + 1);
+				target = target.substr(commaIndex + 1);
+				commaIndex = target.indexOf(',');
+				targetToken = target;
+				if (commaIndex >= 0) {
+					targetToken = target.substr(0, commaIndex);
+					settings = JSON.parse(target.substr(commaIndex + 1));
```
Comment (whalemer) [inline server/chat-commands.js]: Bit of a nitpick but this should be in a try ... catch block and you need to make sure it's an object so you don't crash when using `in` on it.

**Comment by whalemer** [comment]
How do you plan on detecting if a server supports `/trn+`? If you know if a server is updated or not, can't you just include settings based off that?

**Comment by scheibo** [comment]
> How do you plan on detecting if a server supports `/trn+`? If you know if a server is updated or not, can't you just include settings based off that?

Ugh yeah, youre right. Simply adding the new command has its own host of issues -_-. Really dont want to deal with `CAP` negotiation...

**Pull request closed without merging**